src/env/commands/emufs.S

```diff
@@ -47,206 +47,122 @@
+	.word	Strat			# ptr to strategy routine
+	.word	Intr			# ptr to interrupt service routine
+	.ascii	"EMUFS$  "		# logical-device name
 
-Dispatch:
+.align 2
 
-	.word	Init		# initialize driver
-	.word	MediaChk
-	.word	BuildBpb
+RHPtr:	.long	0			# ptr to request header
+
+Dispatch:
+	.word	Init			# initialize driver
 	.word	Dummy
-	.word	Read		# read
 	.word	Dummy
 	.word	Dummy
+	.word	Dummy			# read
 	.word	Dummy
-	.word	Write		# write
-	.word	Write		# write with verify
+	.word	Dummy
+	.word	Dummy
+	.word	Dummy			# write
+	.word	Dummy			# write with verify
 	.word	Dummy
 	.word	Dummy
 	.word	Dummy
 
 Strat:
-# was...
-#	mov	word ptr cs:[RHPtr],bx
-#	mov	word ptr cs:[RHPtr+2],es
 	movw	%bx, %cs:RHPtr
 	movw	%es, %cs:RHPtr+2
 	lret
 
-
 Intr:
-	pushw	%ax
-	pushw	%bx
-	pushw	%cx
-	pushw	%dx
+	pusha
 	pushw	%ds
 	pushw	%es
-	pushw	%di
-	pushw	%si
-	pushw	%bp
-	
+
 	pushw	%cs
 	popw	%ds
 
-	les	RHPtr,%di	# let es:di = request header
+	les	RHPtr,%di		# let es:di = request header
 
-	movb	%es:2(%di),%bl
+	movb	%es:CMD(%di),%bl
 	xorb	%bh,%bh
 	cmpw	$MaxCmd,%bx
-	jle	Intr1
-	call	Error
-	jmp	Intr2
+	jle	1f
+	movw	$0x8003,%ax		# Set error
+	jmp	2f
+
+1:	shlw	%bx
 
-Intr1:	shlw	%bx
-	
 	call	*Dispatch(%bx)
 
 	les	RHPtr,%di
 
-Intr2:	or	$0x100,%ax		# Merge done bit with status
-	movw	%ax,%es:3(%di)
+2:	or	$0x100,%ax		# Merge done bit with status
+	movw	%ax,%es:STATUS(%di)
 
-	popw	%bp
-	popw	%si
-	popw	%di
 	popw	%es
 	popw	%ds
-	popw	%dx
-	popw	%cx
-	popw	%bx
-	popw	%ax
+	popa
 	lret
 
 Dummy:
-	call MFSini
-	les	RHPtr, %di
-	movb	%es:2(%di),%bl
-	movb	$3, %bh
+	push	%es
+	push	%dx
 
-	movw    $DOS_HELPER_MFS_HELPER, %ax
+	push	%cs
+	pop	%es
+	movw	$.LDummyMsg, %dx
+	movw	$DOS_HELPER_PRINT_STRING, %ax
 	int	$DOS_HELPER_INT
 
-	xorw	%ax,%ax
-	ret
-
-
-MediaChk:
-	call MFSini
-# was...
-#	mov	byte ptr es:[di+14],$-1	# disk has changed
-#	mov	word ptr es:[di+15], offset NoName
-#	mov	word ptr es:[di+17], cs	
-	movb	$-1,%es:14(%di)	# disk has changed
-	movw	$NoName,%es:15(%di)
-	movw	%cs,%es:17(%di)
-	xorw	%ax,%ax
-	ret
-
-
-BuildBpb:
-	call MFSini
-# was...
-#	mov	word ptr es:[di+18],offset Bpb
-#	mov	word ptr es:[di+20],cs
-	movw	$Bpb,%es:18(%di)
-	movw	%cs,%es:20(%di)
-	xorw	%ax,%ax
-	ret
-
-
-Read:
-	call MFSini
-	movw	$0x101,%bx
-	movw	$DOS_HELPER_MFS_HELPER,%ax
-	int	$DOS_HELPER_INT
-	xorw	%ax,%ax
-	ret
-
-
-Write:
-	call MFSini
-	movw	$0x202,%bx
-	movw	$DOS_HELPER_MFS_HELPER,%ax
-	int	$DOS_HELPER_INT
-	xorw	%ax,%ax
-	ret
-
+	pop	%dx
+	pop	%es
 
-Error:
+# Tell the caller we don't support this command. For example if we do
+# COPY AUTOEXEC.BAT EMUFS$
+# the user gets to the see the 'Abort, Retry, Fail' message.
 	movw	$0x8003,%ax
 	ret
+.LDummyMsg:
+	.asciz	"emufs: Dummy stub called"
 
-
-MyIret:
-	xorw	%ax,%ax
-	iret
-
+InitCode:	# All Data and code below will be discarded after Init
 
 MFSini:
 	pushw %es
 	pushw %di
 
 	pushw %cs
 	popw %ds
-	movw InitDone,%ax
-	cmp $0,%ax
-	jne AlreadyDone
```

Comment:
Yes.
Also ems.sys can be loaded twice...
Would be nice if you take care of that.
I think the best would be to stuff this
into the version checking. Besides
version OK/old it can also return the
installed status.